src/storage/check.rs

```diff
@@ -1,111 +1,186 @@
-    original: *const Storage<'a, T, D>,
+pub struct CheckStorage<'a, T, D, S> {
+    data: UnsafeCell<RestrictedStorage<'a, T, D, S>>,
 }
 
-impl<'a, 'e, T, D> Join for &'a CheckStorage<'e, T, D> {
-    type Type = Entry<'a, 'e, T, D>;
-    type Value = *const Storage<'e, T, D>;
+impl<'a, T, D, S> Join for &'a CheckStorage<'a, T, D, S>
+    where T: Component,
+          D: Deref<Target = MaskedStorage<T>> + 'a,
+          S: Borrow<Storage<'a, T, D>>,
+{
+    type Type = (Entry<'a, T, D>, &'a RestrictedStorage<'a, T, D, S>);
+    type Value = &'a RestrictedStorage<'a, T, D, S>;
     type Mask = &'a BitSet;
 
     fn open(self) -> (Self::Mask, Self::Value) {
-        (&self.bitset, self.original)
+        let restricted = unsafe { &*self.data.get() };
+        (&restricted.0.borrow().data.mask, restricted)
     }
 
-    unsafe fn get(storage: &mut *const Storage<'e, T, D>, id: Index) -> Entry<'a, 'e, T, D> {
-        Entry {
+    unsafe fn get(storage: &mut Self::Value, id: Index) -> Self::Type {
+        let entry = Entry {
             id: id,
-            original: *storage,
-            phantom: PhantomData,
-        }
+            pointer: storage.0.borrow() as *const Storage<'a, T, D>,
+        };
+        (entry, storage)
     }
 }
 
-unsafe impl<'a, T, D> DistinctStorage for CheckStorage<'a, T, D> {}
+impl<'a, T, D, S> Join for &'a mut CheckStorage<'a, T, D, S>
+    where T: Component,
+          D: DerefMut<Target = MaskedStorage<T>> + 'a,
+          S: BorrowMut<Storage<'a, T, D>>,
+{
+    type Type = (Entry<'a, T, D>, &'a mut RestrictedStorage<'a, T, D, S>);
+    type Value = &'a mut RestrictedStorage<'a, T, D, S>;
+    type Mask = &'a BitSet;
+
+    fn open(self) -> (Self::Mask, Self::Value) {
+        let restricted = unsafe { &*self.data.get() };
+        (&restricted.0.borrow().data.mask, unsafe { &mut *self.data.get() })
+    }
+
+    unsafe fn get(storage: &mut Self::Value, id: Index) -> Self::Type {
+        use std::mem;
+        let storage: &'a mut Self::Value = mem::transmute(storage);
+        let entry = Entry {
+            id: id,
+            pointer: storage.0.borrow() as *const Storage<'a, T, D>,
+        };
+        (entry, storage)
+    }
+}
+
+unsafe impl<'a, T, D, S> DistinctStorage for CheckStorage<'a, T, D, S> { }
+unsafe impl<'a, T, D, S> Send for CheckStorage<'a, T, D, S> { } // UnsafeCell removes thread safety.
+unsafe impl<'a, T, D, S> Sync for CheckStorage<'a, T, D, S> { }
 
 /// An entry to a storage.
-pub struct Entry<'a, 'e, T, D> {
+pub struct Entry<'a, T, D>
+    where T: Component,
+          D: 'a,
+{
     id: Index,
     // Pointer for comparison when attempting to check against a storage.
-    original: *const Storage<'e, T, D>,
-    phantom: PhantomData<&'a ()>,
+    pointer: *const Storage<'a, T, D>,
+}
+
+impl<'a, T, D> Entry<'a, T, D>
+    where T: Component,
+          D: 'a,
+{
+    #[inline]
+    fn assert_same_storage(&self, storage: &Storage<'a, T, D>) {
```

Comment:
Afaik it forces creation of new lifetime for the index and then ensures that that lifetimed index can only be used with right container. You actually need closure for it too. I didn't remember that was the case.